src/ape_ethereum/ecosystem.py

```diff
@@ -147,6 +168,89 @@ def decode_address(cls, raw_address: RawAddress) -> AddressType:
     def encode_address(cls, address: AddressType) -> RawAddress:
         return str(address)
 
+    def is_proxy(self, address: AddressType) -> bool:
+        return self.get_proxy_info(address) is not None
+
+    def get_proxy_info(self, address: AddressType) -> Optional[ProxyInfo]:
+        code = self.provider.get_code(address).hex()[2:]
+        patterns = {
+            ProxyType.Minimal: r"363d3d373d3d3d363d73(.{40})5af43d82803e903d91602b57fd5bf3",
+            ProxyType.Vyper: r"366000600037611000600036600073(.{40})5af4602c57600080fd5b6110006000f3",  # noqa: E501
+            ProxyType.Clones: r"36603057343d52307f830d2d700a97af574b186c80d40429385d24241565b08a7c559ba283a964d9b160203da23d3df35b3d3d3d3d363d3d37363d73(.{40})5af43d3d93803e605b57fd5bf3",  # noqa: E501
+        }
+        for type, pattern in patterns.items():
+            match = re.match(pattern, code)
+            if match:
+                target = self.conversion_manager.convert(match.group(1), AddressType)
+                return ProxyInfo(type=type, target=target)
+
+        keccak = self.provider.web3.keccak
+        slots = {
+            ProxyType.Standard: int(keccak(text="eip1967.proxy.implementation").hex(), 16) - 1,
+            ProxyType.Beacon: int(keccak(text="eip1967.proxy.beacon").hex(), 16) - 1,
+            ProxyType.OpenZeppelin: keccak(text="org.zeppelinos.proxy.implementation"),
+            ProxyType.UUPS: keccak(text="PROXIABLE"),
+        }
+        for type, slot in slots.items():
+            storage = self.provider.web3.eth.get_storage_at(address, slot)
```

Comment:
mypy warns web3 is not guaranteed to be the provider, maybe we should add getting to a provider api